R/tm_g_waterfall.R

```diff
@@ -292,16 +292,85 @@ srv_g_waterfall <- function(id,
-      iv$add_rule("bar_var", shinyvalidate::sv_required())
-      iv$add_rule("bar_paramcd", shinyvalidate::sv_required())
+      iv$add_rule("bar_var", shinyvalidate::sv_required(
+        message = "Bar Height is required"
+      ))
+      iv$add_rule("bar_paramcd", shinyvalidate::sv_required(
+        message = "Tumor Burden Parameter is required"
+      ))
+      iv$add_rule("bar_paramcd", shinyvalidate::sv_in_set(
+        set = adtr$PARAMCD,
+        message_fmt = "All values of Tumor Burden Parameter must be elements of ADTR PARAMCD"
+      ))
+      iv$add_rule("add_label_paramcd_rs", shinyvalidate::sv_optional())
+      iv$add_rule("add_label_paramcd_rs", shinyvalidate::sv_in_set(
+        set = adrs$PARAMCD,
+        message_fmt = "ADRS Label must be an element of ADRS PARAMCD"
+      ))
+      rule_excl <- function(value, other) {
+        if (length(value) > 0L && length(other) > 0L) {
+          "Only one \"Label to Bars\" is allowed"
+        }
+      }
+      iv$add_rule("add_label_paramcd_rs", rule_excl, other = input$add_label_var_sl)
+      iv$add_rule("add_label_var_sl", rule_excl, other = input$add_label_paramcd_rs)
+      iv$add_rule("anno_txt_paramcd_rs", shinyvalidate::sv_optional())
+      iv$add_rule("anno_txt_paramcd_rs", shinyvalidate::sv_in_set(
+        set = adrs$PARAMCD,
+        message_fmt = "Annotation Parameters must be elements of ADRS PARAMCD"
+      ))
+      iv$add_rule("href_line", shinyvalidate::sv_optional())
+      iv$add_rule("href_line", ~ if (anyNA(suppressWarnings(as_numeric_from_comma_sep_str(.)))) {
+        "Horizontal Reference Line(s) are invalid"
+      })
+      iv$add_rule("ytick_at", shinyvalidate::sv_required(
+        message = "Y-axis Interval is required"
+      ))
+      iv$add_rule("ytick_at", ~ if (!checkmate::test_number(suppressWarnings(as.numeric(.)), lower = 1)) {
+        "Y-axis Interval must be a single positive number"
+      })
+      iv$add_rule("gap_point_val", shinyvalidate::sv_optional())
+      iv$add_rule("gap_point_val", ~ if (!checkmate::test_number(suppressWarnings(as.numeric(.)), lower = 1)) {
+        "Break High Bars must be a single positive number"
+      })
       iv$enable()
+      iv
+    })
 
+    output_q <- reactive({
       adsl <- data[["ADSL"]]()
       adtr <- data[[dataname_tr]]()
       adrs <- data[[dataname_rs]]()
 
+      # validate data rows
+      teal::validate_has_data(adsl, min_nrow = 2)
+      teal::validate_has_data(adtr, min_nrow = 2)
+      teal::validate_has_data(adrs, min_nrow = 2)
+
+      adsl_vars <- unique(
+        c(
+          "USUBJID", "STUDYID",
+          input$bar_color_var, input$sort_var, input$add_label_var_sl, input$anno_txt_var_sl, input$facet_var
+        )
+      )
+      adtr_vars <- unique(c("USUBJID", "STUDYID", "PARAMCD", input$bar_var))
+      adrs_vars <- unique(c("USUBJID", "STUDYID", "PARAMCD", "AVALC"))
+      adrs_paramcd <- unique(c(input$add_label_paramcd_rs, input$anno_txt_paramcd_rs))
```

Comment:
`input$bar_var` and `input$anno_txt_paramcd_rs` can be checked in the `iv` reactive

```suggestion
      adtr_vars <- unique(c("USUBJID", "STUDYID", "PARAMCD"))
      adrs_vars <- unique(c("USUBJID", "STUDYID", "PARAMCD", "AVALC"))
```

```r
       iv$add_rule("bar_var", shinyvalidate::sv_in_set(
          names(adtr),
          message_fmt = sprintf("Bar Hight variable must be present in %s", dataname_tr)
        )
       iv$add_rule("add_label_paramcd_rs", shinyvalidate::sv_in_set(
          names(adrs),
          message_fmt = sprintf("Add ADRS Label to Bars variables must be present in %s", dataname_rs)
        )
        iv$add_rule("anno_txt_paramcd_rs", shinyvalidate::sv_in_set(
          names(adrs),
          message_fmt = sprintf("Annotation Parameters variables must be present in %s", dataname_rs)
        )
```